lua/entities/gmod_tardis/modules/sh_tracking.lua

```diff
@@ -0,0 +1,497 @@
+    end
+
+    function ENT:SetTracking(ent,ply)
+        local wasTrackingEnt = self:GetData("tracking-ent")
+        local wasTracking = IsValid(wasTrackingEnt)
+        local isTracking = IsValid(ent)
+        local valid = true
+        if isTracking then 
+            if ent.TardisPart or ent.TardisInterior or (ent:IsPlayer() and IsValid(TARDIS:GetInteriorEnt(ent))) then
+                if IsValid(ply) then
+                    TARDIS:ErrorMessage(ply, "Controls.Tracking.InteriorFail")
+                end
+                valid = false
+            elseif ent == self then
+                if IsValid(ply) then
+                    TARDIS:ErrorMessage(ply, "Controls.Tracking.SelfFail")
+                end
+                valid = false
+            elseif ent:GetPos():Distance(self:GetPos()) > TRACKING_MAX_DISTANCE_SET then
+                if IsValid(ply) then
+                    TARDIS:ErrorMessage(ply, "Controls.Tracking.DistanceFail")
+                end
+                valid = false
+            elseif self:CallHook("CanTrack", ent, ply) == false then
+                if IsValid(ply) then
+                    TARDIS:ErrorMessage(ply, "Controls.Tracking.GenericFail")
+                end
+                valid = false
+            end
+            if not valid then
+                return false
+            end
+
+            local wasFlying = self:GetFlight()
+            if not wasFlying then
+                local success = self:SetFlight(true)
+                if not success then
+                    if IsValid(ply) then
+                        TARDIS:ErrorMessage(ply, "Controls.Tracking.FlightFail")
+                    end
+                    return false
+                end
+            end
+
+            if not wasTracking then
+                self:SetData("tracking-wasflight", wasFlying)
+            end
+        end
+        
+        self:SetData("tracking-ent",ent)
+        if IsValid(ent) and ent ~= wasTrackingEnt then
+            local offsetPos
+            local offsetYaw
+            if self:GetTrackRotation() then
+                offsetPos = ent:WorldToLocal(self:GetPos())
+                offsetYaw = -ent:GetAngles().y + self:GetAngles().y
+            else
+                offsetPos = WorldToLocal(self:GetPos(), angle_zero, ent:GetPos(), angle_zero)
+                offsetYaw = self:GetAngles().y
+            end
+            self:SetData("tracking-offset-pos", offsetPos)
+            self:SetData("tracking-offset-yaw", offsetYaw)
+        end
+
+        if not isTracking then
+            if wasTracking and (not self:GetData("tracking-wasflight")) then
+                self:SetFlight(false)
+            end
+            self:SetData("tracking-wasflight", nil)
+            self:SetData("tracking-offset-pos", nil)
+            self:SetData("tracking-offset-yaw", nil)
+            if IsValid(self.trackingdebugprop) then
+                self.trackingdebugprop:Remove()
+            end
+        end
```

Comment:
Most of the code has `if isTracking` conditions on it, we check for it too many times

Perhaps it would be better to have everything that happens when `not isTracking` to the beginning of the function?

Reply:
Done

https://github.com/MattJeanes/TARDIS/pull/969/commits/7f6800965156a2e84bfba24a6c2b8ec4b3dc7725

Had to duplicate this little bit to make it work but otherwise looks a lot better now

```lua
if wasTracking ~= isTracking then
    TARDIS:StatusMessage(ply, "Controls.Tracking.Status", isTracking)
end
```